#585 Constant value initialization with static expressions

sergisiso opened the pull request:
This PR makes the DataSymbol constant_values as PSyIR nodes in order to allow static expressions.
The change in the front-end and back-end is relatively simple, it just needs to continue recusing the `process_nodes`/`visit`.

But constant values can also be set by the DataSymbol setter. In this case it still allows for both, PSyIR expressions and Python intrinsic types which will be converted to PSyIR Literals
e.g.
```
sym = DataSymbol('a', DataType.REAL, constant_value=3.1415)
sym = DataSymbol('a', DataType.INTEGER, constant_value=BinaryOperation(...))
```

sergisiso commented:
@arporter @rupertford This is ready for the first review.

arporter commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -670,17 +670,11 @@ def process_declarations(self, parent, nodes, arg_list):
-                        if isinstance(expr, Fortran2003.NumberBase):
-                            value_str = expr.items[0]
-                            # Convert string literal to the Symbol datatype
-                            ct_value = TYPE_MAP_TO_PYTHON[datatype](value_str)
```
pylint now says that TYPE_MAP_TO_PYTHON is unused in this module.

sergisiso replied:
Deleted

arporter commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -342,21 +342,24 @@ def unresolved_interface(self):
-        instance.
-
-        '''
+        :param constant_value: Set or change the fixed known value of the \
```
s/constant_value/new_value/ same for the the :type: below.

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -368,21 +371,39 @@ def constant_value(self, new_value):
+            if isinstance(new_value, Node):
+                for node in new_value.walk(Node):
+                    if not isinstance(node, (Literal, Operation, Reference)):
+                        raise ValueError(
```
Should this be a NotImplementedError? i.e. have we chosen to just support a sub-set of what is possible?

sergisiso replied:
Response in a comment below

arporter commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -368,21 +371,39 @@ def constant_value(self, new_value):
+                        "instance datatype is '{1}' which means the constant "
+                        "value is expected to be '{2}' but found '{3}'."
+                        "".format(self.name, self.datatype,
+                                  TYPE_MAP_TO_PYTHON[self.datatype],
```
You could use the `lookup` variable here as you've already done this dict lookup at line 386.

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -368,21 +371,39 @@ def constant_value(self, new_value):
+                                  TYPE_MAP_TO_PYTHON[self.datatype],
+                                  type(new_value)))
+
+                # All accepted types but booleans need to be passed as strings
```
s/but/other than/

sergisiso replied:
Deleted this sentence altogether as booleans are also strings now

arporter commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -368,21 +371,39 @@ def constant_value(self, new_value):
+
+                # All accepted types but booleans need to be passed as strings
+                if isinstance(new_value, bool):
+                    new_value = Literal(new_value, self.datatype)
```
So parameter values are now stored as `Literal` objects instead of just the raw value?

sergisiso replied:
Yes, parameters are now PSyIR expressions and it fits better if they are Literals in the case of single values.

arporter commented on src/psyclone/tests/psyir/frontend/fparser2_test.py:
```diff
@@ -343,22 +343,24 @@ def test_process_declarations(f2008_parser):
 
     # Static constant expresions are not supported
-    reader = FortranStringReader("real, parameter :: a = 1.1, b = a * 2")
+    reader = FortranStringReader("real, parameter :: i4 = 1.1, i5 = i4 * 2")
```
You need to update the comment above this line :-)

sergisiso replied:
Ups.. done

arporter commented on src/psyclone/tests/psyir/frontend/fparser2_test.py:
```diff
@@ -565,7 +567,20 @@ def test_process_declarations_kind_new_param(f2008_parser):
 
-@pytest.mark.xfail(reason="Kind parameter declarations not supported - #585")
+def test_process_declarations_expression_param(f2008_parser):
+    ''' Test that process_declarations handles the kind attribute when
```
There's no kind parameter in the Fortran being used in this test?

sergisiso replied:
Note that this xfail is for the `test_process_declarations_kind_param` test and I updated the issue number because it still fails because the KIND() intrinsic is still not recognized.

arporter replied:
Thanks. I think I must have got confused.

arporter commented on src/psyclone/tests/psyir/frontend/fparser2_test.py:
```diff
@@ -565,7 +567,20 @@ def test_process_declarations_kind_new_param(f2008_parser):
+    '''
+    fake_parent = KernelSchedule("dummy_schedule")
+    processor = Fparser2Reader()
+    reader = FortranStringReader("real, parameter :: twopi = 2 * pi")
```
Isn't this a repeat of the test at line 358?

sergisiso replied:
True. I removed this test now.

arporter commented on src/psyclone/tests/psyir/symbols/datasymbol_test.py:
```diff
@@ -250,23 +251,42 @@ def test_datasymbol_constant_value_setter():
+    lhs = Literal('2', DataType.INTEGER)
+    rhs = Reference('constval')
+    ct_expr = BinaryOperation.create(BinaryOperation.Operator.ADD, lhs, rhs)
+    sym = DataSymbol('a', DataType.INTEGER, constant_value=ct_expr)
```
Please could you check that the newly-created DataSymbol has the expected properties.

arporter replied:
Also, the docstring for the DataSymbol constructor needs updating now that `constant_value` can be PSyIR.

sergisiso replied:
Added asserts and updated the DataSymbol constructor docstring.

arporter commented on src/psyclone/tests/psyir/symbols/datasymbol_test.py:
```diff
@@ -250,23 +251,42 @@ def test_datasymbol_constant_value_setter():
+    # Test with invalid constant expressions
+    ct_expr = Return()
+    with pytest.raises(ValueError) as error:
+        sym = DataSymbol('a', DataType.INTEGER, constant_value=ct_expr)
```
In this case, contrary to what I said earlier, I think a ValueError is the most appropriate. Are there cases where it should be a NotImplementedError?

sergisiso replied:
I think in this context ValueError makes more sense as it is a PSyIR symbol and we define what it is supported or not in the PSyIR.

If there is some Fortran initialization expression that fails this check, we could catch the error in the fparser2reader and raise a GenerationError or NotImplementedError then. That probably is more correct but I am unsure to add this error catch&rename expressions because then the code stops in a different part of the code when it breaks. What do you think?

sergisiso commented:
@arporter This PR is ready for the next review. I addressed the inline comments and as suggested I also updated the Literal class so that booleans are stored as strings like the other types. Then the proper formatting for Fortran booleans need to be managed by the frontend/backend which in turn solves #616

sergisiso commented on src/psyclone/tests/psyir/frontend/literals_test.py:
```diff
@@ -67,29 +67,22 @@ def test_handling_literal(f2008_parser, code, dtype):
+    processor = Fparser2Reader()
+    with pytest.raises(GenerationError) as error:
+        processor.process_nodes(fake_parent, [astmt], None)
+    assert "Exptected to find '.true.' or '.false' as fparser2 logical " \
```
fixed

sergisiso commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -45,10 +45,10 @@
     Schedule, Directive, CodeBlock, IfBlock, Reference, Literal, Loop, \
     KernelSchedule, Container, Assignment, Return, Array, InternalError, \
-    GenerationError
+    GenerationError, Node, Operation
```
deleted

sergisiso commented on src/psyclone/tests/psyir/frontend/fparser2_test.py:
```diff
@@ -565,7 +567,7 @@ def test_process_declarations_kind_new_param(f2008_parser):
 
 
-@pytest.mark.xfail(reason="Kind parameter declarations not supported - #585")
+@pytest.mark.xfail(reason="Kind parameter declarations not supported - #569")
```
Done (and also did the same to all other tests in this file that don't need the fixture object)

sergisiso commented on src/psyclone/tests/psyir/frontend/literals_test.py:
```diff
@@ -67,29 +67,22 @@ def test_handling_literal(f2008_parser, code, dtype):
-        Literal(True, DataType.INTEGER)
-    assert ("non-boolean Literal must be supplied with a value encoded as a "
-            "string but got" in str(err.value))
+def test_handling_invalid_logic_literal(f2008_parser):
```
Done

sergisiso commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -3632,10 +3630,42 @@ def test_array_create_invalid():
 # Test Literal class
-def test_literal_value():
-    '''Test the value property returns the value of the Literal object.
+def test_literal_init():
```
Done

sergisiso commented on src/psyclone/psyir/symbols/datasymbol.py:
```diff
@@ -368,22 +372,38 @@ def constant_value(self, new_value):
+                        "'{3}'.".format(self.name, self.datatype, lookup,
+                                        type(new_value)))
+                if self.datatype == DataType.BOOLEAN:
+                    new_value = str(new_value).lower()
```
We already check it in the previous `isinstance(new_value, lookup)` condition, however I see this was not very clear and I refactored and added comments in this section. In the new version I do not redefine `new_value` as it was confusing.

arporter replied:
Ah, I understand now. Thanks for tidying/explaining.

sergisiso commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -3632,10 +3630,42 @@ def test_array_create_invalid():
+    # Test invalid boolean value
+    with pytest.raises(ValueError) as err:
+        Literal("invalid", DataType.BOOLEAN)
+    assert "A DataType.BOOLEAN Literal can only be: 'true' or 'false' " \
```
I added the suggested tests in both `test_literal_init` and `test_literal_init_invalid`.

Regarding trailing/leading white space I would say it shouldn't be part of Literal as it is the frontend responsibility to deal with this.

However, I am unsure about the case sensitivity, I see pros/cons for each alternative. I currently left it just accepting lower case (and the front-ends need to  conform to this), but I don't mind changing it.

arporter replied:
Oh yes, I was getting confused and forgetting that this is internal. I agree that we should leave it the way you have it now.

sergisiso commented:
@arporter Ready for next review

arporter commented on src/psyclone/tests/psyir/nodes/literal_test.py:
```diff
@@ -0,0 +1,114 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2017-2019, Science and Technology Facilities Council.
```
I think this should be 2019-2020
